plugins/out_datadog/datadog.c

```diff
@@ -0,0 +1,279 @@
+                                          tag, sizeof(tag_len)-1);
+        }
+
+        /* dd_source */
+        if (ctx->dd_source != NULL) {
+            dd_msgpack_pack_key_value_str(&mp_pck,
+                                          FLB_DATADOG_DD_SOURCE_KEY, sizeof(FLB_DATADOG_DD_SOURCE_KEY) -1,
+                                          ctx->dd_source, flb_sds_len(ctx->dd_source));
+        }
+
+        /* dd_service */
+        if (ctx->dd_service != NULL) {
+            dd_msgpack_pack_key_value_str(&mp_pck,
+                                          FLB_DATADOG_DD_SERVICE_KEY, sizeof(FLB_DATADOG_DD_SERVICE_KEY) -1,
+                                          ctx->dd_service, flb_sds_len(ctx->dd_service));
+        }
+
+        /* dd_tags */
+        if (ctx->dd_tags != NULL) {
+            dd_msgpack_pack_key_value_str(&mp_pck,
+                                          FLB_DATADOG_DD_TAGS_KEY, sizeof(FLB_DATADOG_DD_TAGS_KEY) -1,
+                                          ctx->dd_tags, flb_sds_len(ctx->dd_tags));
+        }
+
+        /* Append initial object k/v */
+        int i = 0;
+        for (i = 0; i < map_size; i++) {
+            k = map.via.map.ptr[i].key;
+            v = map.via.map.ptr[i].val;
+
+            msgpack_pack_object(&mp_pck, k);
+            msgpack_pack_object(&mp_pck, v);
+        }
+    }
+
+    /* Convert from msgpack to JSON */
+    out_buf = flb_msgpack_raw_to_json_sds(mp_sbuf.data, mp_sbuf.size);
+    msgpack_sbuffer_destroy(&mp_sbuf);
+
+    if (!out_buf) {
+        flb_error("[out_datadog] error formatting JSON payload");
+        msgpack_unpacked_destroy(&result);
+        return -1;
+    }
+
+    *out_data = out_buf;
+    *out_size = flb_sds_len(out_buf);
+    /* Cleanup */
+    msgpack_unpacked_destroy(&result);
+
+    return 0;
+}
+
+static void cb_datadog_flush(const void *data, size_t bytes,
+                             const char *tag, int tag_len,
+                             struct flb_input_instance *i_ins,
+                             void *out_context,
+                             struct flb_config *config)
+{
+    struct flb_out_datadog *ctx = out_context;
+    struct flb_upstream_conn *upstream_conn;
+    struct flb_http_client *client;
+
+    flb_sds_t payload_buf;
+    size_t payload_size = 0;
+    size_t b_sent;
+    int ret = FLB_ERROR;
+
+    /* Get upstream connection */
+    upstream_conn = flb_upstream_conn_get(ctx->upstream);
+    if (!upstream_conn) {
+        FLB_OUTPUT_RETURN(FLB_RETRY);
+    }
+
+    /* Convert input data into a Datadog JSON payload */
+    ret = datadog_format(data, bytes, tag, tag_len, &payload_buf, &payload_size, ctx);
+    if (ret == -1) {
+        FLB_OUTPUT_RETURN(FLB_ERROR);
+    }
+
+    /* Create HTTP client context */
+    client = flb_http_client(upstream_conn, FLB_HTTP_POST, ctx->uri,
+                             payload_buf, payload_size,
+                             ctx->host, ctx->port,
+                             NULL, 0);
+
```

Comment:
always check if client is not null, if so, release resources allocated.

Reply:
Yes, saw it after fixing your first comment :)